src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -416,90 +343,117 @@ private BoundExpression RewriteStringConcatInExpressionLambda(SyntaxNode syntax,
+                    var toStringMethod = (MethodSymbol)member;
+                    if (toStringMethod.GetLeastOverriddenMethod(type) == (object)objectToStringMethod)
+                    {
+                        structToStringMethod = toStringMethod;
+                        break;
                     }
                 }
             }
 
-            // Optimization not possible; just return the original expression.
-            return expr;
-        }
+            // If it's a special value type, it should have its own ToString method (but we might fail to find
+            // it if object.ToString is missing).  Assume that this won't be removed, and emit a direct call rather
+            // than a constrained virtual call. This keeps in the spirit of #7079, but expands the range of
+            // types to all special value types.
+            if (structToStringMethod != null && expr.Type.SpecialType != SpecialType.None)
+            {
+                return BoundCall.Synthesized(expr.Syntax, expr, structToStringMethod);
+            }
 
-        /// <summary>
-        /// Gets whether the type of an argument used in string concatenation can
-        /// be optimized by first calling ToString on it before passing the argument
-        /// to the String.Concat function.
-        /// </summary>
-        /// <param name="symbol">The type symbol of the argument.</param>
-        /// <returns>
-        /// true if ToString may be used; false if using ToString could lead to observable differences in behavior.
-        /// </returns>
-        private static bool ConcatExprCanBeOptimizedWithToString(TypeSymbol symbol)
-        {
-            // There are several constraints applied here in support of backwards compatibility:
-            // - This optimization potentially changes the order in which ToString is called
-            //   on the arguments.  That's a a compatibility issue if one argument's ToString
-            //   depends on state mutated by another, such as current culture.
-            // - For value types, this optimization causes ToString to be called on the original
-            //   value rather than on a boxed copy.  That means a mutating ToString implementation
-            //   could change the original rather than the copy.
-            // For these reasons, this optimization is currently restricted to primitives
-            // known to have a non-mutating ToString implementation that is independent
-            // of externally mutable state.  Common value types such as Int32 and Double
-            // do not meet this bar.
-
-            switch (symbol.SpecialType)
+            // - It's a reference type (excluding unconstrained generics): no copy
+            // - It's a constant: no copy
+            // - The type definitely doesn't have its own ToString method (i.e. we're definitely calling 
+            //   object.ToString on a struct type, not a generic type): no copy (yes this is a versioning issue,
```

Comment:
I don't think I understand what you mean by "calling object.ToString on a struct type, not a generic type". When you say "generic type" do you mean "type parameter"?

Reply:
I do mean that: I'll fix the comment thanks!